service/history/queueProcessor.go

```diff
@@ -0,0 +1,361 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber-common/bark"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	// QueueProcessorOptions is options passed to queue processor implementation
+	QueueProcessorOptions struct {
+		BatchSize           int
+		WorkerCount         int
+		MaxPollRPS          int
+		MaxPollInterval     time.Duration
+		UpdateAckInterval   time.Duration
+		ForceUpdateInterval time.Duration
+		MetricScope         int
+	}
+
+	queueProcessorBase struct {
+		shard         ShardContext
+		options       *QueueProcessorOptions
+		processor     processor
+		logger        bark.Logger
+		metricsClient metrics.Client
+		rateLimiter   common.TokenBucket // Read rate limiter
+		ackMgr        *ackManager
+
+		notifyCh   chan struct{}
+		isStarted  int32
+		isStopped  int32
+		shutdownWG sync.WaitGroup
+		shutdownCh chan struct{}
+	}
+
+	// ackManager is created by QueueProcessor to keep track of the queue ackLevel for the shard.
+	// It keeps track of read level when dispatching tasks to processor and maintains a map of outstanding tasks.
+	// Outstanding tasks map uses the task id sequencer as the key, which is used by updateAckLevel to move the ack level
+	// for the shard when all preceding tasks are acknowledged.
+	ackManager struct {
```

Comment:
can we separate this ack manager into a dedicated file? that will be easier to add unit test and have separation of logic which will make the code more readable

Reply:
For now it is not used by any other component and only implementation detail of queue processor.  Let me put a comment to move this outside.